RationaleMCP/0031/concept.md

```diff
@@ -0,0 +1,135 @@
+   * prefixes
+   * attributes of variables 
+   * information lost in preceding transformation steps,
+   * information to control further transformation steps,
+   * information to support the numerical solution of the equation system
+   * source code location information, 
+   * other data for special use cases, e.g. packaging into a FMU, inputs and outputs of the whole model.
+* Description strings (also called syntactic comments) are an exception. These are optionally copied as such.
+* An external tool does not need any knowledge of special Modelica semantics, e.g. handling start attributes, behavior of relations and built-in operators for events,etc..
+
+## Details
+
+* All higher level syntactical and structural contructs of Modelica are removed, such as:
+   * imports
+   * inheritance
+   * redeclarations
+   * nested class definitions (There is no class definition inside another class definition)
+   * packages, [expandable] connectors, models (only one model remains), blocks, operator definitions
+   * functional inputs of functions
+   * conditional declarations
+   * modifications (therefore no merging of them)
+   * connect statements
+   * loops inside equation sections
+   * `if`/`then`/`else` in equation sections (atleast with different sized branches)
+   * `when` clauses
+   * stream operators: `actualStream(...)`, `inStream(...)`
+   * all prefixes, except `input` and `output`
+* __TODO__ synchronous Modelica extensions, including state machines, possibly introduction of partitions
+
+### Names
+
+* Names are defined as the `IDENT` token in Modelica
+* The hierarchy of the Modelica model is encoded into the names of the variables and types by separation of the levels by a dot.
+* Names from Modelica are generally enclosed in single quotation marks, whether or not they contain a dot.  
+* Generated helper variables do not need single quotes. 
+* __TODO__: array components, combining names for helper variables
+
+### Types
+
+* Following types exist:
+   * predefined types: `Real`, `Integer`, `Boolean`, `String`, `ExternalObject`,
+   * record types
+   * functions
+   * simple types.
+   
+* Properties of record definitions are:
+   * The components are defined as variable definitions. Declaration equations are not allowed.
+   * There are no equations or algorithms.
+   * record definitions can not nest.
+   * There is no automatism to generate record constructors anymore. They have to be made explicit as functions.
+
+* Simple types are aliases to predefined types enriched with annotations. 
+   * They are used to avoid duplication of annotations in variable definitions. 
+   * An annotation of a variable definition overrides the same annotation of the type.
+   * They have no dimension.
+
+* The order of definition is not important.
+
+### Variables
+
+* A variable definition consists of
+   1. a prefix (e.g. `input`,`output`),
+   1. a type name,
+   1. a variable name
+   1. an optional dimension specification,
```

Comment:
I'd like the dimension specification together with the rest of the type.

Reply:
I agree that dimensions should only be at one place, but whether dimensions are at the variable definitions or at the type is a matter of taste I would say.
If it is at the variable definition there is a need to duplicate dimension information for each instance. If it is at the type definition, there is a need to generate artificial types if the dimension was originally at the variable.
Also the handling vectorization of complex component models would hard if the dimensions are at the type, because there are no such types in the language anymore, because it is flat. 
Therefore dimensions at types somehow violate the requirement of the language to be flat and I would keep them at the variables.

Reply:
I didn't mean to introduce type aliases, I just meant to use the form
```
Real[2] 'x';
```
instead of
```
Real x[2];
```

Of course, the form that would make most sense when creating arrays of arrays would be to say
```
type Arr = [3, 4] Real;
[2] Real 'x'; /* 'x' has type [2, 3, 4] Real  */
```
but as we have no precedence for this order in Modelica, I don't think it's worth pursuing the idea.